pkg/syncer/cnsoperator/controller/csinodetopology/csinodetopology_controller.go

```diff
@@ -343,16 +345,62 @@ func getNodeTopologyInfo(ctx context.Context, nodeVM *cnsvsphere.VirtualMachine,
 	log.Infof("NodeVM %q belongs to topology: %+v", nodeVM.Reference(), topologyCategories)
 	topologyLabels := make([]csinodetopologyv1alpha1.TopologyLabel, 0)
 	if isZoneRegion {
-		// In order to keep backward compatibility, we will use the kubernetes
-		// standard topology labels for zone and region.
+		// Create the kubernetes client.
+		k8sClient, err := k8s.NewClient(ctx)
+		if err != nil {
+			log.Errorf("failed to create K8s client. Error: %v", err)
+			return nil, err
+		}
+
+		// In order to keep backward compatibility, we will discover existing topology labels
+		// on nodes in the cluster. Few examples on how the discovery works:
+		// Node    Topology label
+		// N1      topology.kubernetes.io/XYZ
+		// N2      topology.kubernetes.io/XYZ
+		// After the cluster is up and running, N1 and N2 will have topology labels of the
+		// form `topology.kubernetes.io/XYZ`
+		//
+		// Node    Topology label
+		// N1      failure-domain.beta.kubernetes.io/XYZ
+		// N2      No label
+		// After the cluster is up and running, N1 and N2 will have topology labels of the
+		// form `failure-domain.beta.kubernetes.io/XYZ`
+		//
+		// Node    Topology label
+		// N1      failure-domain.beta.kubernetes.io/XYZ
+		// N2      topology.kubernetes.io/XYZ
+		// The nodes in the cluster will remain in CrashLoopBackOff state till the customer
+		// intervenes to have a uniform label across all nodes in the cluster.
+		//
+		// Node    Topology label
+		// N1      No label
+		// N2      No label
+		// After the cluster is up and running, N1 and N2 will have topology labels of the
+		// form `topology.csi.vmware.com/XYZ`
+		nodeList, err := k8sClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return nil, logger.LogNewErrorf(log, "failed to list nodes in the cluster. Error: %+v", err)
+		}
+		zoneLabel, regionLabel, err := findExistingTopologyLabels(ctx, nodeList.Items)
+		if err != nil {
+			return nil, logger.LogNewErrorf(log, "failed to discover existing topology labels "+
+				"on the nodes in the cluster. Error: %+v", err)
+		}
+		if zoneLabel == "" && regionLabel == "" {
```

Comment:
Would it make more sense if this part of the code(Line:389 - Line:394) is moved inside  `findExistingTopologyLabels` method ?

Reply:
I thought of doing that at first but then the function name says `finding existing topology labels` which should ideally be empty if GA and beta labels are not found. That is why I moved this code out.